src/blockrelay/mempool_sync.cpp

```diff
@@ -0,0 +1,503 @@
+                nRemainingMempoolBytes -= nTxSize;
+            }
+        }
+
+        if (mempoolTxHashes.size() == 0)
+        {
+            LOG(MPOOLSYNC, "Mempool is empty; aborting mempool sync with peer %s\n", pfrom->GetLogName());
+            return true;
+        }
+
+        // Assemble mempool sync object
+        uint64_t nBothMempools = mempoolTxHashes.size() + mempoolinfo.nTxInMempool;
+        CMempoolSync mempoolSync(mempoolTxHashes, mempoolinfo.nTxInMempool, nBothMempools, mempoolinfo.shorttxidk0,
+            mempoolinfo.shorttxidk1, NegotiateMempoolSyncVersion(pfrom));
+
+        pfrom->PushMessage(NetMsgType::MEMPOOLSYNC, mempoolSync);
+        LOG(MPOOLSYNC, "Sent mempool sync to peer %s using version %d\n", pfrom->GetLogName(), mempoolSync.version);
+    }
+    else
+    {
+        dosMan.Misbehaving(pfrom, 100);
+
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Handle an incoming mempool synchronization payload
+ */
+bool CMempoolSync::ReceiveMempoolSync(CDataStream &vRecv, CNode *pfrom, std::string strCommand)
+{
+    // Deserialize mempool sync payload
+    CMempoolSync tmp;
+    vRecv >> tmp;
+    auto mempoolSync = std::make_shared<CMempoolSync>(std::forward<CMempoolSync>(tmp));
+
+    LOG(MPOOLSYNC, "Received mempool sync from peer %s\n", pfrom->GetLogName());
+
+    // Do not process unrequested mempool sync.
+    {
+        LOCK(cs_mempoolsync);
+
+        if (!(mempoolSyncRequested.count(pfrom) == 1))
+        {
+            dosMan.Misbehaving(pfrom, 10);
+            return error("Received unrequested mempool sync from peer %s\n", pfrom->GetLogName());
+        }
+
+        // Do not proceed if this request has already been processed
+        if (mempoolSyncRequested[pfrom].completed)
+        {
+            dosMan.Misbehaving(pfrom, 100);
+            return error(
+                "Received mempool sync from peer %s but synchronization has already completed", pfrom->GetLogName());
+        }
+    }
+
+    return mempoolSync->process(pfrom);
+}
+
+bool CMempoolSync::process(CNode *pfrom)
+{
+    std::set<uint256> passingTxHashes;
+    std::map<uint64_t, uint256> mapPartialTxHash;
+    std::set<uint64_t> setHashesToRequest;
+
+    std::vector<uint256> mempoolTxHashes;
+    GetMempoolTxHashes(mempoolTxHashes);
+
+    // Collect cheap hashes
+    {
+        LOCK(cs_mempoolsync);
+
+        for (const uint256 &hash : mempoolTxHashes)
+        {
+            uint64_t cheapHash = GetShortID(mempoolSyncRequested[pfrom].shorttxidk0,
+                mempoolSyncRequested[pfrom].shorttxidk1, hash, SHORT_ID_VERSION);
+            mapPartialTxHash.insert(std::make_pair(cheapHash, hash));
+        }
+    }
+
+    try
+    {
+        std::vector<uint64_t> blockCheapHashes = pGrapheneSet->Reconcile(mapPartialTxHash);
+
+        // Sort out what hashes we have from the complete set of cheapHashes
+        for (size_t i = 0; i < blockCheapHashes.size(); i++)
+        {
```

Comment:
nit: this could be a range based loop
-for (size_t i = 0; i < blockCheapHashes.size(); i++)
-uint64_t cheapHash = blockCheapHashes[i];
+for (uint64_t cheapHash : blockCheapHahes)